{"cells":[{"attachments":{},"cell_type":"markdown","metadata":{},"source":["> Remember to create a branch named `exceptions`\n","\n","# Exceptions\n","\n","In Python, errors in the code are named \"Exceptions\". Said exceptions are, as expected, descendant classes of a base class, `Exception`. Its variants include, amongst others:\n","\n","- `ZeroDivisionError`: generated when dividing by `0`.\n","- `IndexError`: generated due to attempted access to indices over the total length of a sequential structure.\n","- `NameError`: generated due to a variable, function or class call that has not been defined previously.\n","- `SyntaxError`: generated due to the incorrect spelling of a keyword or the usage of unsupported symbols."]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["## Traceback\n","\n","Traceback (from the words \"trace\" and \"back\") is exactly what you expect: a detailed retrospective over how and where the exception was generated. They usually offer relevant information, but can easily turn misleading when the traceback is too long."]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["## Exception handling\n","\n","Exception handling is the technique used to detect and handle exceptions. The handling process aims to redirect the execution flow to a \"safe\" alteranative, instead of just letting the exception be raised."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Without exception catch:\n","\n","def divide(value_1, value_2):\n","    return value_1 / value_2\n","\n","\n","print(divide(4, 2), end=\"\\n\\n\")\n","print(divide(2, 0))\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# With exception catch:\n","\n","def divide(value_1, value_2):\n","    output = None\n","\n","    try:\n","        output = value_1 / value_2\n","    except ZeroDivisionError:\n","        print(\"Warning: The second value cannot be zero.\")\n","\n","    return output\n","\n","\n","print(divide(4, 2), end=\"\\n\\n\")\n","print(divide(2, 0))\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Exception catch with else statement:\n","\n","def divide(value_1, value_2):\n","    output = None\n","\n","    try:\n","        output = value_1 / value_2\n","    except ZeroDivisionError:\n","        print(\"Warning: The second value cannot be zero.\")\n","    else:\n","        print(\"The calculation was OK.\")\n","\n","    return output\n","\n","\n","print(divide(4, 2), end=\"\\n\\n\")\n","print(divide(2, 0))\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Exception catch with else statement:\n","\n","def divide(value_1, value_2):\n","    output = None\n","\n","    try:\n","        output = value_1 / value_2\n","    except ZeroDivisionError:\n","        print(\"Warning: The second value cannot be zero.\")\n","    else:\n","        print(\"The calculation was OK.\")\n","    finally:\n","        print(\"Finished calculation process.\")\n","\n","    return output\n","\n","\n","print(divide(4, 2), end=\"\\n\\n\")\n","print(divide(2, 0))\n"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["## Exception raising\n","\n","Exception raising is sort of the opposite version of exception handling. By raising an exception, you are causing the code to fail the moment it reaches the instruction. This can be really useful when dealing with data type enforcements (i.e. an user passes a `str` to your function that is only intended to receive `int`s, then an exception is raised)."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Exception raising:\n","\n","def sum_squared(*args):\n","    for arg in args:\n","        if not isinstance(arg, (int, float)):\n","            raise TypeError(\"all values must be integers or floats.\")\n","\n","    return sum(arg ** 2 for arg in args)\n","\n","print(\"No exception raised: \", sum_squared(1, 2, 3))\n","print(\"Exception raised: \", sum_squared(1, 2, \"c\"))\n"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["## _Exercise 25: Exception management_\n","\n","Steps:\n","\n","1. Define a function named `weird_number` that accepts a single positional argument.\n","    1. If the argument is not an integer, float or string, raise a `TypeError` with the message `\"this function only accepts integers, floats and strings.\"`.\n","    2. **Try** to cast the argument to a float. If it does not work, set the value of the argument to `0`. If it works, set the value of the argument to its cube (power).\n","    3. **Finally**, add one to the value of the argument.\n","    4. Return the value.\n","\n","- [Click here to open the script in the editor](./exercises/exercise_25.py)\n","- Test the script using `Ctrl + Shift + P` > `Tasks: Run Task` > `Test exercise`"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["> Remember to create a pull request for branch `exceptions`\n","\n","# Navigation\n","\n","- **Previous lesson**: [Classes](../classes/theory.ipynb)\n","- **Next lesson**: [Libraries](../libraries/theory.ipynb)"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyOqkdzlgvaxPWf/B3qgPJlZ","collapsed_sections":["s65z3Wt39b7W","fLWkiKzx9fTm","qzhCjBRBzRDa","-DKcF6q50gAp","pgjDQlN605j5","a0ISIYYt_iq3","neF9KusEw2Dy","PS8rpsCx7fyA","sacz-IDM7tmw","-Kkpfwh68bGA","lStQYj0E8mF3","3WT73f5s8xkQ","ZJ_C0uyA87XA","JGAZ-AUq915N","4uVom-gr-ZCA","g_qBK6LM-jqH","gYoN-W2N3shP","_GliAtBS5Zaq","MsOUWERC9NLm","qJ8rUYRN6NlC","l_-ET-iN89al","rR2ie8Ke7xwy","fcTmCxzH834B","Biie78Sq_EbJ","Hunm3Sma_rbd","jchnU4E2QvQS","fgJ-M-EiRg9q","xxNSgsRaR_SC","0B4SwWJXUl7N","lCHmecQ5VJCV","KHXI_Od6UcN5","5QpkFs3xbS2Z","QjjeOOf8d96B"],"provenance":[{"file_id":"1jH2xcSVbi6rEtypaoTc0huKAgzPFuln1","timestamp":1666127789028}]},"kernelspec":{"display_name":".venv","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.1"},"vscode":{"interpreter":{"hash":"edb3f8401898146638121a13b0247bbb0e9c0cce851d580080f4cb1d49fc2b60"}}},"nbformat":4,"nbformat_minor":0}
